src/filter/json_encode_filter.rs

```diff
@@ -0,0 +1,268 @@
+        results[0] // fail if results empty, else return processed event
+    }
+
+    #[test]
+    fn parsable_line_parsing_off() {
+        // Test we don't parse a line if parsing is off
+        assert_eq!(
+            process_event(false, metric::Event::LogLine {
+                path: "testpath",
+                value: "{\"bad\": \"do not parse\"}",
+                time: 946684800,
+                tags: Default::default(),
+                fields: Default::default(),
+            }),
+            json!({
+                "path": "testpath",
+                "value": "{\"bad\": \"do not parse\"}",
+                "time": "2000-01-01T00:00:00",
+                "tags": {},
+            })
+        );
+    }
+
+    #[test]
+    fn parsable_line_parsing_on() {
+        // Test we do parse a line if parsing is on
+        assert_eq!(
+            process_event(true, metric::Event::LogLine {
+                path: "testpath",
+                value: "{\"good\": \"do parse\"}",
+                time: 946684800,
+                tags: Default::default(),
+                fields: Default::default(),
+            }),
+            json!({
+                "path": "testpath",
+                "good": "do parse",
+                "time": "2000-01-01T00:00:00",
+                "tags": {},
+            })
+        );
+    }
+
+    #[test]
+    fn unparsable_line() {
+        // Test we don't parse a line if it's not JSON
+        assert_eq!(
+            process_event(true, metric::Event::LogLine {
+                path: "testpath",
+                value: "this is not json",
+                time: 946684800,
+                tags: Default::default(),
+                fields: Default::default(),
+            }),
+            json!({
+                "path": "testpath",
+                "value": "this is not json",
+                "time": "2000-01-01T00:00:00",
+                "tags": {},
+            })
+        );
+    }
+
+    #[test]
+    fn non_object_line() {
+        // Test we don't parse a line if it's not a JSON object but is valid JSON
+        assert_eq!(
+            process_event(true, metric::Event::LogLine {
+                path: "testpath",
+                value: "[123, \"not an object\"]",
+                time: 946684800,
+                tags: Default::default(),
+                fields: Default::default(),
+            }),
+            json!({
+                "path": "testpath",
+                "value": "[123, \"not an object\"]",
+                "time": "2000-01-01T00:00:00",
+                "tags": {},
+            })
+        );
+    }
+
+    // quickcheck and serde_json::map::Map aren't compatible, so we ask quickcheck for
+    // many Vec<(String, String)>s that we turn into maps.
+    fn vecs_to_objs(vecs: &Vec<Vec<(String, String)>>) -> Vec<Map<String, Value>> {
+        iter(vecs).map(|vec|
+            iter(vec).map(|(k, v)| (k.clone(), v.clone().into()))
```

Comment:
It doesn't matter in-test, but all the cloning here could be avoided if `vecs_to_objs` consumed `vecs`. Right now you're passing by reference and then cloning data from inside the clone.

Reply:
yeah. I was able to figure that out, but couldn't figure out how to make a consuming iterator such that `k` and `v` would be `String` instead of `&String`.

Reply:
Try `into_iter`.